diag_manager/include/fms_diag_fieldbuff_update.fh

```diff
@@ -0,0 +1,1336 @@
+!***********************************************************************
+!*                   GNU Lesser General Public License
+!*
+!* This file is part of the GFDL Flexible Modeling System (FMS).
+!*
+!* FMS is free software: you can redistribute it and/or modify it under
+!* the terms of the GNU Lesser General Public License as published by
+!* the Free Software Foundation, either version 3 of the License, or (at
+!* your option) any later version.
+!*
+!* FMS is distributed in the hope that it will be useful, but WITHOUT
+!* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+!* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+!* for more details.
+!*
+!* You should have received a copy of the GNU Lesser General Public
+!* License along with FMS.  If not, see <http://www.gnu.org/licenses/>.
+!***********************************************************************
+
+ !> @brief This code will be used by the preprecessor to generate an implementation
+ !! to the module procudure for the fieldbuff_update interface. The
+  !! generated function is a wrapper calling 4D field/5D buffer version of the same.
+FUNCTION FMS_DIAG_FBU_3D_PNAME_(ofield_cfg, ofield_index_cfg, field_d, sample, &
+  & ofb, ofc, bbounds, count_0d, num_elements, mask, weight1, missvalue,  &
+  & field_num_threads, field_active_omp_level, issued_mask_ignore_warning, &
+  & l_start, l_end, err_msg,  err_msg_local ) result( succeded )
```

Comment:
Aren't a lot of these arguments optional?  I think at least some of them should be optional.  Maybe I'm wrong.

Reply:
There are a few different  things:  1) Weight is weight1 in send_data_3d, and not the optional "weight".
2) Concerning mask, it is is optional in send_data_3d, but right now when it is not provided by the user, we are passing a dummy allocated var to this routine. It was a workaround to a pointer bounds remapping. I may have fixed the issue  but well see. If the issue is gone, then that case mask will be put back as optional. Also, line 
73ff already has this
```
IF  (PRESENT (mask)) THEN
      mask_ptr(1:size(mask,1),1:size(mask,2),1:size(mask,3),1:1) => mask
    ENDIF
```
(BTW, whether or not to use the mask mask comes from the flag "mask_present" that is in fmsDiagOutfield_typefmsDiagOutfield_type. Obviously id the mask is not present in send_data_3d, then the flag needs to be false.)
3) err_msg is optional.
4) ofc may or may not be allocated when passed in, so there is a check in line 66.